# 1.

In [150]:
import sys
import numpy as np
import jupyterlab
import time
import random
np.set_printoptions(threshold=20)

1. **КЛАССИЧЕСКИЙ ПИТОН**

In [151]:
t_range = 100000
products = ['автомобиль HAVAL', 'обувь', 'одежда', 'телефоны']
sales = [random.choices(range(0,101), k=t_range) for _ in range(len(products))]
prices = [20000, 3000, 45600, 5000]

In [152]:
s = [0,0,0,0]

In [212]:
start=time.time()
for i in range(len(products)):
    for j in range(t_range):
        s[i]=s[i]+prices[i]*sales[i][j]
end=time.time()
print('TIME:', end-start)

TIME: 0.07873392105102539


Сумма, вырчученная за продажи HAVAL, обуви, одежды и телефонов соответственно:

In [154]:
s

[100176380000, 14990037000, 228698500800, 25038015000]

2. **NUMPY** 

In [155]:
products_n = np.array(['автомобиль HAVAL', 'обувь', 'одежда', 'телефоны'])
prices_n = np.array([20000, 3000, 45600, 5000])

In [156]:
sales_n=np.array(sales, dtype="uint64")
sales_n

array([[  8,  15,  23, ...,  40,  55, 100],
       [ 99,  63,  77, ...,  25,  74,  14],
       [ 68,  62,  77, ...,  98,  99,  16],
       [ 89,   0,  20, ...,  49,  98,  86]], dtype=uint64)

In [157]:
start=time.time()
s_n=sales_n.sum(axis=1)*prices
end=time.time()
print('TIME:', end-start)

TIME: 0.0009984970092773438


Сумма, вырчученная за продажи HAVAL, обуви, одежды и телефонов соответственно:

In [158]:
s_n

array([1.00176380e+11, 1.49900370e+10, 2.28698501e+11, 2.50380150e+10])

3. **NUMPY  с векторизацией**

In [211]:
s_m=np.zeros((1,t_range), dtype=float)
start=time.time()
for i in range (len(products)):
    print(np.sum(np.multiply(sales_n[i], prices[i], out=s_m)))
end=time.time()
print('TIME:', end-start)

100176380000.0
14990037000.0
228698500800.0
25038015000.0
TIME: 0.001992940902709961


#### **ИТОГИ**

В результате все данные подсчетов сошлись.
Сравним время выполнения:

1. 0,078 секунд
2. 0,001 секунда - numpy справился с задачей заметтно быстрее
3. 0,002 секунда - моя векторная реализация оказалась медленней встроенных инструментов numpy.


# 2.

1. **NUMPY SORT**

In [214]:
x=np.random.randint(1,101,10000)
x

array([53, 71, 69, ..., 95, 11, 18])

In [228]:
%%timeit
np.sort(x, kind='heapsort')

605 µs ± 12.5 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [229]:
print(np.sort(x, kind='heapsort'))

[  1   1   1 ... 100 100 100]


2. **КЛАССИЧЕСКИЙ PYTHON**


In [163]:
def heap_data(nums, index, heap_size):
    largest_num = index
    left_index = 2 * index + 1
    right_index = 2 * index + 2
    if left_index < heap_size and nums[left_index] > nums[largest_num]:
        largest_num = left_index

    if right_index < heap_size and nums[right_index] > nums[largest_num]:
        largest_num = right_index
    if largest_num != index:
        nums[largest_num], nums[index] = nums[index], nums[largest_num]
        heap_data(nums, largest_num, heap_size)
def heap_sort(nums):
    n = len(nums)
    for i in range(n // 2 - 1, -1, -1):
        heap_data(nums, i, n)
    for i in range(n - 1, 0, -1):
        nums[0], nums[i] = nums[i], nums[0]
        heap_data(nums, 0, i)
    return nums

In [164]:
x=np.random.randint(1,101,10000)
x

array([40, 41, 77, ..., 83, 31, 72])

In [165]:
%%timeit
heap_sort(x)

91.3 ms ± 1.5 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [166]:
x

array([  1,   1,   1, ..., 100, 100, 100])

#### **ИТОГИ**

Сортировка numpy - 605 µs = 0.000605 секунд

Сортировка классикой - 91.3 мс = 0.0913 секунд

# 3.

In [167]:
data = np.genfromtxt('Z:\\WB\\2_Py\\drinks.csv', delimiter=',', skip_header=1, dtype={
        "names": ["country", "beer", "spirit", "wine", "total"],
        "formats": ["<S50", "<i4","<i4","<i4","<f8"]
    })


Топ-10 стран, пьющих пиво:

In [230]:
arr_beer=np.sort(data, order = ['beer', 'total'])[::-1]
arr_beer[:10][['country', 'beer']]


array([(b'Namibia', 376), (b'Czech Republic', 361), (b'Gabon', 347),
       (b'Germany', 346), (b'Lithuania', 343), (b'Poland', 343),
       (b'Venezuela', 333), (b'Ireland', 313), (b'Palau', 306),
       (b'Romania', 297)],
      dtype={'names': ['country', 'beer'], 'formats': ['S50', '<i4'], 'offsets': [0, 50], 'itemsize': 70})

Топ-10 стран, пьющих вино:

In [169]:
arr_wine=np.sort(data, order = ['wine', 'total'])[::-1]
arr_wine[:10][['country', 'wine']]

array([(b'France', 370), (b'Portugal', 339), (b'Andorra', 312),
       (b'Switzerland', 280), (b'Denmark', 278), (b'Slovenia', 276),
       (b'Luxembourg', 271), (b'Croatia', 254), (b'Italy', 237),
       (b'Equatorial Guinea', 233)],
      dtype={'names': ['country', 'wine'], 'formats': ['S50', '<i4'], 'offsets': [0, 58], 'itemsize': 70})

Топ-10 пьющих стран:

In [170]:
arr_total=np.sort(data, order = ['total'])[::-1]
arr_total[:10][['country', 'total']]

array([(b'Belarus', 14.4), (b'Lithuania', 12.9), (b'Andorra', 12.4),
       (b'Grenada', 11.9), (b'France', 11.8), (b'Czech Republic', 11.8),
       (b'Russian Federation', 11.5), (b'Slovakia', 11.4),
       (b'Luxembourg', 11.4), (b'Ireland', 11.4)],
      dtype={'names': ['country', 'total'], 'formats': ['S50', '<f8'], 'offsets': [0, 62], 'itemsize': 70})

Топ самых трезвых стран:

In [235]:
y=(np.sort(data[data['total']<=2], order = ['total']))
y

array([(b'Afghanistan',   0,  0, 0, 0.), (b'Bangladesh',   0,  0, 0, 0.),
       (b'Iran',   0,  0, 0, 0.), ..., (b'Bahrain',  42, 63, 7, 2.),
       (b'Fiji',  77, 35, 1, 2.), (b'Vietnam', 111,  2, 1, 2.)],
      dtype=[('country', 'S50'), ('beer', '<i4'), ('spirit', '<i4'), ('wine', '<i4'), ('total', '<f8')])

# **4**

1.

In [173]:
df = np.genfromtxt('Z:\\WB\\2_Py\\tarantino.csv', delimiter=',', skip_header=1, dtype={
        "names": ["movie", "type", "word",	"minutes_in"],
        "formats": ["<S50", "<S50","<S50","<f8"]
    })

#### Заведем df_w, в котором будем хранить только часть df с ругательствами.

In [174]:
df_w = df[df['type']== b'word']
df_w

array([(b'Reservoir Dogs', b'word', b'dick',   0.4 ),
       (b'Reservoir Dogs', b'word', b'dicks',   0.43),
       (b'Reservoir Dogs', b'word', b'fucked',   0.55), ...,
       (b'Jackie Brown', b'word', b'ass', 142.43),
       (b'Jackie Brown', b'word', b'fucking', 142.47),
       (b'Jackie Brown', b'word', b'goddamn', 142.97)],
      dtype=[('movie', 'S50'), ('type', 'S50'), ('word', 'S50'), ('minutes_in', '<f8')])

#### Заведем df_d, в котором будем хранить только часть df с кровопролитием.

In [175]:
df_d = df[df['word']== b'']
df_d

array([(b'Reservoir Dogs', b'death', b'',  59.93),
       (b'Reservoir Dogs', b'death', b'',  85.58),
       (b'Reservoir Dogs', b'death', b'',  85.58), ...,
       (b'Jackie Brown', b'death', b'', 113.28),
       (b'Jackie Brown', b'death', b'', 125.43),
       (b'Jackie Brown', b'death', b'', 143.13)],
      dtype=[('movie', 'S50'), ('type', 'S50'), ('word', 'S50'), ('minutes_in', '<f8')])

#### Зведем массив movies со всеми названиями фильмов.

In [176]:
movies = np.unique(df['movie'])
movies

array([b'Django Unchained', b'Inglorious Basterds', b'Jackie Brown',
       b'Kill Bill: Vol. 1', b'Kill Bill: Vol. 2', b'Pulp Fiction',
       b'Reservoir Dogs'], dtype='|S50')

#### ... и всеми ругательствами.

In [177]:
swr = np.unique(df_w['word'])
swr

array([b'ass', b'asses', b'asshead', ..., b'slut', b'squaw', b'wetback'],
      dtype='|S50')

#### Посчитаем, сколько ругательств и кровопролитий в базе:

In [179]:
np.count_nonzero(df['word'])

1704

In [238]:
np.count_nonzero(df)-np.count_nonzero(df['word'])

190

#### Выведем количество ругательст и кровопролитий в каждом фильме:

In [240]:
for i in range (len(movies)):
    print(movies[i], '-------',
          np.count_nonzero(df_w ['movie']==movies[i]), '-------', 
          100*np.count_nonzero(df_w ['movie']==movies[i])/np.count_nonzero(df['word']))  

b'Django Unchained' ------- 262 ------- 15.375586854460094
b'Inglorious Basterds' ------- 58 ------- 3.403755868544601
b'Jackie Brown' ------- 368 ------- 21.5962441314554
b'Kill Bill: Vol. 1' ------- 57 ------- 3.3450704225352115
b'Kill Bill: Vol. 2' ------- 69 ------- 4.049295774647887
b'Pulp Fiction' ------- 469 ------- 27.523474178403756
b'Reservoir Dogs' ------- 421 ------- 24.706572769953052


In [241]:
for i in range (len(movies)):
    print(movies[i], '-------',
          np.count_nonzero(df_d ['movie']==movies[i]), '-------',
          100*np.count_nonzero(df_d ['movie']==movies[i])/(np.count_nonzero(df)-np.count_nonzero(df['word'])))

b'Django Unchained' ------- 47 ------- 24.736842105263158
b'Inglorious Basterds' ------- 48 ------- 25.263157894736842
b'Jackie Brown' ------- 4 ------- 2.1052631578947367
b'Kill Bill: Vol. 1' ------- 63 ------- 33.1578947368421
b'Kill Bill: Vol. 2' ------- 11 ------- 5.7894736842105265
b'Pulp Fiction' ------- 7 ------- 3.6842105263157894
b'Reservoir Dogs' ------- 10 ------- 5.2631578947368425


2.

#### Выведем количество ругательств в каждом фильме:

In [242]:
for i in range (len(swr)):
    print(swr[i], '-------',
          np.count_nonzero(df_w ['word']==swr[i]), '-------',
          100*np.count_nonzero(df_w ['word']==swr[i])/np.count_nonzero(df['word']))  

b'ass' ------- 140 ------- 8.215962441314554
b'asses' ------- 4 ------- 0.2347417840375587
b'asshead' ------- 1 ------- 0.05868544600938967
b'asshole' ------- 10 ------- 0.5868544600938967
b'assholes' ------- 5 ------- 0.2934272300469484
b'bastard' ------- 10 ------- 0.5868544600938967
b'bastards' ------- 1 ------- 0.05868544600938967
b'bitch' ------- 64 ------- 3.755868544600939
b'bitches' ------- 9 ------- 0.528169014084507
b'bullshit' ------- 15 ------- 0.8802816901408451
b'chickenshit' ------- 1 ------- 0.05868544600938967
b'cockblockery' ------- 1 ------- 0.05868544600938967
b'cocksucker' ------- 2 ------- 0.11737089201877934
b'cunt' ------- 2 ------- 0.11737089201877934
b'cunts' ------- 1 ------- 0.05868544600938967
b'damn' ------- 37 ------- 2.171361502347418
b'damned' ------- 3 ------- 0.176056338028169
b'dick' ------- 14 ------- 0.8215962441314554
b'dickless' ------- 1 ------- 0.05868544600938967
b'dicks' ------- 5 ------- 0.2934272300469484
b'dumbass' ------- 1 ------- 0.0586

3.

In [185]:
df_w_ = np.zeros(
    np.count_nonzero(df_w),
    dtype={
        "names": ["movie", "word", "minutes_in"],
        "formats": ["U16", "U16", "f8"]
    }
)
df_w_

array([('', '', 0.), ('', '', 0.), ('', '', 0.), ..., ('', '', 0.),
       ('', '', 0.), ('', '', 0.)],
      dtype=[('movie', '<U16'), ('word', '<U16'), ('minutes_in', '<f8')])

In [197]:
df_w_["movie"] = df_w["movie"]
df_w_["word"] = df_w["word"]
df_w_["minutes_in"] = df[df["type"]==b'word']["minutes_in"]

df_w_

array([('Reservoir Dogs', 'dick',   0.4 ),
       ('Reservoir Dogs', 'dicks',   0.43),
       ('Reservoir Dogs', 'fucked',   0.55), ...,
       ('Jackie Brown', 'ass', 142.43),
       ('Jackie Brown', 'fucking', 142.47),
       ('Jackie Brown', 'goddamn', 142.97)],
      dtype=[('movie', '<U16'), ('word', '<U16'), ('minutes_in', '<f8')])

#### Вычтем из столбца с таймкодами ругательств его же, но сдвинутыый на 1 позицию

In [198]:
diff=np.zeros(len(df_w))
diff=df_w_["minutes_in"][1::]-df_w_["minutes_in"][:-1:]
diff

array([0.03, 0.12, 0.06, ..., 0.5 , 0.04, 0.5 ])

In [199]:
diff=np.insert(diff, 0, df_w_["minutes_in"][0])
diff

array([0.4 , 0.03, 0.12, ..., 0.5 , 0.04, 0.5 ])

In [200]:
df_w_["minutes_in"]=diff
df_w_[["movie","minutes_in"]]

array([('Reservoir Dogs', 0.4 ), ('Reservoir Dogs', 0.03),
       ('Reservoir Dogs', 0.12), ..., ('Jackie Brown', 0.5 ),
       ('Jackie Brown', 0.04), ('Jackie Brown', 0.5 )],
      dtype={'names': ['movie', 'minutes_in'], 'formats': ['<U16', '<f8'], 'offsets': [0, 128], 'itemsize': 136})

#### Какой процент ругательств нужно ждать более минуты

In [209]:
100*np.count_nonzero(df_w_["minutes_in"]>1)/np.count_nonzero(df_w_)


11.150234741784038

#### ...более 5 минут

In [210]:
100*np.count_nonzero(df_w_["minutes_in"]>5)/np.count_nonzero(df_w_)


1.2910798122065728

#### Какое ругательство приходится ждать дольше всего и в каком фильме

In [245]:
df_w_[df_w_["minutes_in"]==np.max(df_w_["minutes_in"])]

array([('Kill Bill: Vol. ', 'fucking', 16.72)],
      dtype=[('movie', '<U16'), ('word', '<U16'), ('minutes_in', '<f8')])